charms/interface_tokens/provides.py

```diff
@@ -0,0 +1,70 @@
+import json
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from ops import CharmBase, Relation, Unit
+
+from .model import RequiresModel
+
+log = logging.getLogger("TokensProvider")
+
+
+@dataclass
+class Request:
+    """Represents a request for an authentication token.
+
+    Attributes:
+        unit (str): The name of the unit making the request.
+        relation_id (int): The ID of the relation associated with the request.
+        request (Dict[str,str]): Mapping of the token requests consisting of user:group.
+    """
+
+    unit: str
+    relation_id: int
+    requests: Dict[str, str]
+
+
+class TokensProvider:
+    """Implements the Provides side of the tokens interface."""
+
+    def __init__(self, charm: CharmBase, endpoint: str):
+        self.charm = charm
+        self.endpoint = endpoint
+
+    def send_token(self, request: Request, tokens: Dict[str, str]) -> None:
+        """Send the token to the related units."""
+        for relation in self.relations:
```

Comment:
does this need the same treatment of

```suggestion
        for relation in self.relations or []:
```

or should the `relations` property just return a `List[Relation]`